Log periodically predictions on train set to Tensorboard

thomasv-k1 opened the pull request.

howard-mahe commented on tools/visual_utils/konboi_visualizer.py:
```diff
@@ -0,0 +1,186 @@
+        self.intrinsics = intrinsics
+        self.object_mapping = ObjectMapping()
+
+    def sensors_transforms(self) -> dict[str, np.ndarray]:
```
method naming lacks a verb e.g. `get`

howard-mahe commented on tools/visual_utils/konboi_visualizer.py:
```diff
@@ -0,0 +1,186 @@
+        boxes: list[list[float]],
+        categories: list[str],
+    ):
+        transforms = self.sensors_transforms()
```
this method could run once and for all in __init__

howard-mahe commented on tools/visual_utils/konboi_visualizer.py:
```diff
@@ -0,0 +1,186 @@
+
+        cam_points = points_homogeneous @ T_lidar_to_cam.T  # (N, 8, 4)
+
+        return cam_points[:, :, :3]  # shape (N, 8, 3)
```
general tip: it is quite sometimes quite convenient to use `[..., :3]` notation

howard-mahe commented on tools/visual_utils/konboi_visualizer.py:
```diff
@@ -0,0 +1,186 @@
+        return sensors
+
+    def transform_lidar_to_camera(self, points: np.ndarray, T_lidar_to_cam: np.ndarray) -> np.ndarray:
+        # points: (N, 8, 3)
```
add docstring with expected shape

howard-mahe commented on tools/visual_utils/konboi_visualizer.py:
```diff
@@ -0,0 +1,186 @@
+
+        return cam_points[:, :, :3]  # shape (N, 8, 3)
+
+    def project_points_to_image(self, points_3d: np.ndarray, K: np.ndarray) -> np.ndarray:
```
add docstring with expected shape

howard-mahe commented on tools/visual_utils/konboi_visualizer.py:
```diff
@@ -0,0 +1,186 @@
+
+    def is_box_visible_in_camera(self, bbox: np.ndarray) -> bool:
+        # Check if all corners are visible from camera (Z > 0)
+        return np.all(bbox[:, 2] > 0)
```
method could be static.

necessary but not sufficient condition. check that points lie within image boundaries would make it sufficient.
i don't mind if we do not perform the camera frustum check as long as it works, but in that case `is_in_front_of_camera` sounds more accurate

howard-mahe commented on tools/train_utils/train_utils.py:
```diff
@@ -1,14 +1,82 @@
+
+
+def load_prediction_tensorboard(model, batch, epoch, tb_log):
+    def load_images(scene_dir, id):
```
`load_images` and `load_sensors_calibration` functions are super dataset-dependent and should not really be here.

would it possible for `load_prediction_tensorboard` to get the dataset object (from `train_loader.dataset` typically), then check that a method `get_related_vision_data(index)` exists and if so, proceed to log model predictions on related images ?

howard-mahe commented on tools/train_utils/train_utils.py:
```diff
@@ -1,14 +1,82 @@
+
+        return intrinsics, extrinsics
+
+    data_path = Path("/home/thomas/konboi/OpenPCDet/data/custom")
```
hardcoded path. one more reason, to count on `Dataset` object to handle that

howard-mahe commented on tools/train_utils/train_utils.py:
```diff
@@ -1,14 +1,82 @@
+    return first_scene
+
+
+def load_prediction_tensorboard(model, batch, epoch, tb_log):
```
add docstring

thomasv-k1 commented:
This commit was made prior to reviewing your feedback.

It includes two main updates:

- Refactored the logic for plotting predictions and added support for composing a single image combining front, left, and right views in `plot_utils.py`.
- Added prediction visualizations to TensorBoard during evaluation.

Any points you've raised that aren't yet addressed in this commit will be handled in the following ones.
